services/consensuscontext/create_block.go

```diff
@@ -19,21 +22,62 @@ func (s *service) createTransactionsBlock(ctx context.Context, blockHeight primi
 	}
 	txCount := len(proposedTransactions.SignedTransactions)
 
+	merkleTransactionsRoot, err := CalculateTransactionsRootHash(proposedTransactions.SignedTransactions)
+	if err != nil {
+		return nil, err
+	}
+
 	txBlock := &protocol.TransactionsBlockContainer{
 		Header: (&protocol.TransactionsBlockHeaderBuilder{
-			ProtocolVersion: primitives.ProtocolVersion(1), // TODO: fix
-			BlockHeight:     blockHeight,
-			//Timestamp: 			   primitives.TimestampNano(time.Now().UnixNano()),
+			ProtocolVersion:       primitives.ProtocolVersion(s.config.ProtocolVersion()),
+			VirtualChainId:        s.config.VirtualChainId(),
+			BlockHeight:           blockHeight,
 			PrevBlockHashPtr:      prevBlockHash,
+			Timestamp:             primitives.TimestampNano(time.Now().UnixNano()),
+			TransactionsRootHash:  merkleTransactionsRoot,
+			MetadataHash:          nil,
 			NumSignedTransactions: uint32(txCount),
 		}).Build(),
 		Metadata:           (&protocol.TransactionsBlockMetadataBuilder{}).Build(),
 		SignedTransactions: proposedTransactions.SignedTransactions,
 		BlockProof:         nil,
 	}
+
+	s.metrics.transactionsRate.Measure(int64(len(txBlock.SignedTransactions)))
 	return txBlock, nil
 }
 
+func CalculateTransactionsRootHash(txs []*protocol.SignedTransaction) (primitives.MerkleSha256, error) {
+	forest, root := merkle.NewForest()
+	diffs := make([]*merkle.MerkleDiff, len(txs))
+	for i := 0; i < len(txs); i++ {
```

Comment:
why not use `range`?
```suggestion
	for i, tx := range txs {
```

Reply:
right, changed it here and other places in the file